### PR TITLE
feat: add creating directory through context menu in navigation tree

### LINKED ISSUE #915: Add create folder action to database/folder context menu
1. only relative path
2. name with slashes `dir2/dir3` creates a tree of directories
3. full path of the directory being created is displayed as i type

### PR DESCRIPTION
**Pull request by walborn**
closes #915

### CODE REVIEW

**Comment by Raubzeug** [comment]
Press "Enter" does't confirm creating a directory.

File: src/containers/Tenant/i18n/en.json [hunk 1]
```diff
@@ -36,5 +36,11 @@
   "actions.alterTable": "Alter table...",
   "actions.alterTopic": "Alter topic...",
   "actions.selectQuery": "Select query...",
-  "actions.upsertQuery": "Upsert query..."
+  "actions.upsertQuery": "Upsert query...",
+  "actions.createDirectory": "Create directory",
+  "schema.tree.dialog.placeholder": "Relative path",
+  "schema.tree.dialog.header": "Create directory",
+  "schema.tree.dialog.description": "Inside directory",
```
Comment (Raubzeug) [inline src/containers/Tenant/i18n/en.json:43]: As for me, just "Inside" will be enough. Otherwise we get "Create directory inside directory".

File: src/containers/Tenant/Schema/SchemaTree/SchemaTree.tsx [hunk 2]
```diff
@@ -57,26 +70,70 @@ export function SchemaTree(props: SchemaTreeProps) {
...
-            renderAdditionalNodeElements={getControls(dispatch, {
-                setActivePath: onActivePathUpdate,
-            })}
-            activePath={currentPath}
-            onActivePathUpdate={onActivePathUpdate}
-            cache={false}
-            virtualize
-        />
+        <React.Fragment>
+            <Dialog open={open} onClose={handleCreateDirectoryClose}>
```
Comment (Raubzeug) [inline src/containers/Tenant/Schema/SchemaTree/SchemaTree.tsx:93]: Let's move dialog to separate component

File: src/containers/Tenant/Schema/SchemaTree/SchemaTree.tsx [hunk 3]
```diff
@@ -19,10 +28,14 @@ interface SchemaTreeProps {
...
     const {rootPath, rootName, rootType, currentPath, onActivePathUpdate} = props;
-
     const dispatch = useTypedDispatch();
 
     const [_, setQueryMode] = useQueryModes();
+    const [open, setOpen] = React.useState(false);
+    const [parent, setParent] = React.useState<string>('');
+    const [child, setChild] = React.useState('');
+
+    const [createDirectory, {requestId}] = schemaApi.useCreateDirectoryMutation();
```
Comment (Raubzeug) [inline src/containers/Tenant/Schema/SchemaTree/SchemaTree.tsx:38]: Are you sure we should use `requestId` as a key to re-render?
If request fails, we don't have to repaint.
And when we get this id? What if we have slow network, and requestId is already here, but directory isn't created yet?

File: src/containers/Tenant/Schema/SchemaTree/SchemaTree.scss [hunk 4]
```diff
@@ -0,0 +1,14 @@
+@import '../../../../styles/mixins.scss';
+
+.ydb-schema-tree {
+    &__modal {
```
Comment (Raubzeug) [inline src/containers/Tenant/Schema/SchemaTree/SchemaTree.scss:4]: do we really need this class?

File: src/containers/Tenant/Schema/SchemaTree/SchemaTree.scss [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+        max-height: 100%;
+        @include flex-container();
+    }
+    &__label {
+        margin-bottom: 10px;
+    }
+    &__description {
+        opacity: 0.6;
```
Comment (Raubzeug) [inline src/containers/Tenant/Schema/SchemaTree/SchemaTree.scss:12]: Lets better use `color: var(--g-color-text-secondary)`

File: src/containers/Tenant/Schema/SchemaTree/SchemaTree.tsx [hunk 6]
```diff
@@ -57,26 +70,70 @@ export function SchemaTree(props: SchemaTreeProps) {
...
+            <Dialog open={open} onClose={handleCreateDirectoryClose}>
+                <Dialog.Header caption={i18n('schema.tree.dialog.header')} />
+                <Dialog.Body className={b('modal')}>
+                    <div className={b('label')}>
+                        <div className={b('description')}>
+                            {i18n('schema.tree.dialog.description')}
+                        </div>
+                        <div>{`${parent}/`}</div>
+                    </div>
+                    <TextInput
```
Comment (Raubzeug) [inline src/containers/Tenant/Schema/SchemaTree/SchemaTree.tsx:102]: Could you please clarify with Xeno if we need any validation here?

**Review by Raubzeug — CHANGES_REQUESTED** [review]

File: src/containers/Tenant/Schema/SchemaTree/SchemaTree.tsx [hunk 7]
```diff
@@ -50,33 +58,66 @@ export function SchemaTree(props: SchemaTreeProps) {
...
     }, [currentPath, onActivePathUpdate, rootPath]);
 
+    const handleCreateDirectoryClose = () => {
+        setCreateDirectoryOpen(false);
+    };
+
+    const handleCreateDirectorySubmit = (child: string) => {
+        createDirectory({database: parent, path: `${parent}/${child}`});
+
+        onActivePathUpdate(`${parent}/${child}`);
```
Comment (Raubzeug) [inline src/containers/Tenant/Schema/SchemaTree/SchemaTree.tsx:81]: It seems we should do it after directory is created (wait for `createDirectory` answer).

File: src/containers/Tenant/Schema/SchemaTree/SchemaTree.tsx [hunk 8]
```diff
@@ -50,33 +58,66 @@ export function SchemaTree(props: SchemaTreeProps) {
...
+
+        setCreateDirectoryOpen(false);
```
Comment (Raubzeug) [inline src/containers/Tenant/Schema/SchemaTree/SchemaTree.tsx:83]: same here. I would set some `loading` state to dialog `Apply` buttons and close dialog on request success. If request fails we may show an error inside the dialog's body.

File: src/containers/Tenant/Schema/CreateDirectoryDialog/CreateDirectoryDialog.scss [hunk 9]
```diff
@@ -0,0 +1,14 @@
+@import '../../../../styles/mixins.scss';
+
+.ydb-schema-create-directory-dialog {
+    &__modal {
```
Comment (Raubzeug) [inline src/containers/Tenant/Schema/CreateDirectoryDialog/CreateDirectoryDialog.scss:4]: It seems this class is redundant.

**Review by Raubzeug — CHANGES_REQUESTED** [review]

**Comment by Raubzeug** [comment]
One more - we should disable "Apply" button, if directory name is empty.

File: src/containers/Tenant/Schema/SchemaTree/SchemaTree.tsx [hunk 10]
```diff
@@ -50,33 +59,80 @@ export function SchemaTree(props: SchemaTreeProps) {
         return childItems;
     };
 
+    React.useEffect(() => {
+        if (createDirectoryResult.isSuccess) {
+            setNavigationTreeKey(createDirectoryResult.requestId);
+            onActivePathUpdate(`${parent}/${child}`);
+            setCreateDirectoryOpen(false);
+        }
+    }, [createDirectoryResult, onActivePathUpdate, parent, child]);
```
Comment (ValeraS) [inline src/containers/Tenant/Schema/SchemaTree/SchemaTree.tsx:68]: `createDirectory` returns promise, we can make these actions on promise resolve.

File: src/containers/Tenant/Schema/CreateDirectoryDialog/CreateDirectoryDialog.tsx [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+                <TextInput
+                    placeholder={i18n('schema.tree.dialog.placeholder')}
+                    value={child}
+                    onUpdate={handleUpdate}
+                    autoFocus
+                    hasClear
+                    disabled={isLoading}
+                    validationState={error || invalid ? 'invalid' : undefined}
+                />
+                {error && <div className={b('error')}>{error}</div>}
```
Comment (ValeraS) [inline src/containers/Tenant/Schema/CreateDirectoryDialog/CreateDirectoryDialog.tsx:68]: There is a `ResponseError` component for this.

File: src/containers/Tenant/Schema/CreateDirectoryDialog/CreateDirectoryDialog.tsx [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+
+    React.useEffect(() => {
+        if (!open) {
+            setChild('');
+        }
+    }, [open]);
+
+    React.useEffect(() => {
+        onUpdate(child);
+    }, [onUpdate, child]);
```
Comment (ValeraS) [inline src/containers/Tenant/Schema/CreateDirectoryDialog/CreateDirectoryDialog.tsx:45]: Why is this here and not in `handleSubmit`?

File: src/containers/Tenant/Schema/CreateDirectoryDialog/CreateDirectoryDialog.tsx [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+
+    const handleUpdate = (value: string) => {
+        setChild(value);
+    };
+
+    React.useEffect(() => {
+        if (!open) {
+            setChild('');
+        }
+    }, [open]);
```
Comment (ValeraS) [inline src/containers/Tenant/Schema/CreateDirectoryDialog/CreateDirectoryDialog.tsx:41]: It should be in `handleClose`

File: src/containers/Tenant/Schema/SchemaTree/SchemaTree.tsx [hunk 14]
```diff
@@ -50,33 +59,80 @@ export function SchemaTree(props: SchemaTreeProps) {
...
+
+    React.useEffect(() => {
+        if (createDirectoryResult.isError) {
+            const errorMessage =
+                (createDirectoryResult.error as {data: string})?.data || ' Unknown error';
+            setError(errorMessage);
+        }
+    }, [createDirectoryResult]);
```
Comment (ValeraS) [inline src/containers/Tenant/Schema/SchemaTree/SchemaTree.tsx:76]: Why do use error state? Error already presents in `createDirectoryResult`.

File: src/containers/Tenant/Schema/SchemaTree/SchemaTree.tsx [hunk 15]
```diff
@@ -19,19 +23,24 @@ interface SchemaTreeProps {
...
-
     const dispatch = useTypedDispatch();
 
     const [_, setQueryMode] = useQueryModes();
+    const [createDirectoryOpen, setCreateDirectoryOpen] = React.useState(false);
+    const [parent, setParent] = React.useState<string>('');
+    const [child, setChild] = React.useState<string>('');
+    const [navigationTreeKey, setNavigationTreeKey] = React.useState('');
+    const [error, setError] = React.useState<string>('');
+    const [createDirectory, createDirectoryResult] = schemaApi.useCreateDirectoryMutation();
```
Comment (ValeraS) [inline src/containers/Tenant/Schema/SchemaTree/SchemaTree.tsx:34]: Seems, this should be inside `CreateDirectoryDialog`.

File: src/containers/Tenant/Schema/SchemaTree/SchemaTree.tsx [hunk 16]
```diff
@@ -50,33 +59,80 @@ export function SchemaTree(props: SchemaTreeProps) {
...
+            setNavigationTreeKey(createDirectoryResult.requestId);
+            onActivePathUpdate(`${parent}/${child}`);
+            setCreateDirectoryOpen(false);
+        }
+    }, [createDirectoryResult, onActivePathUpdate, parent, child]);
+
+    React.useEffect(() => {
+        if (createDirectoryResult.isError) {
+            const errorMessage =
+                (createDirectoryResult.error as {data: string})?.data || ' Unknown error';
```
Comment (ValeraS) [inline src/containers/Tenant/Schema/SchemaTree/SchemaTree.tsx:73]: Does the error contain a `message`?

**Review by ValeraS — CHANGES_REQUESTED** [review]
Please try not using UseEffect, all current uses are redundant.

File: src/containers/Tenant/Schema/SchemaTree/SchemaTree.tsx (hunk 16 above)
Comment (walborn) [reply src/containers/Tenant/Schema/SchemaTree/SchemaTree.tsx]: The error contains only such fields
`{
    "data": "{ status: SCHEME_ERROR, issues: { <main>: Error: Check failed: path: '/local/.sys_health/±', error: symbol '?' is not allowed in the path part '±' } }",
    "status": 400,
    "statusText": "Bad Request",
    "headers": {
        "content-length": "140",
        "content-type": "text/plain"
    },
    "config": {
        "transitional": {
            "silentJSONParsing": true,
            "forcedJSONParsing": true,
            "clarifyTimeoutError": false
        },
        "adapter": [
            "xhr",
            "http"
        ],
        "transformRequest": [
            null
        ],
        "transformResponse": [
            null
        ],
        "timeout": 60000,
        "xsrfCookieName": "",
        "xsrfHeaderName": "X-XSRF-TOKEN",
        "maxContentLength": -1,
        "maxBodyLength": -1,
        "env": {},
        "headers": {
            "Accept": "application/json, text/plain, */*",
            "Content-Type": "application/json"
        },
        "withCredentials": false,
        "signal": {},
        "method": "post",
        "url": "http://localhost:8765/scheme/directory",
        "data": "{}",
        "params": {
            "database": "/local/.sys_health",
            "path": "/local/.sys_health/±"
        },
        "axios-retry": {
            "retries": 3,
            "shouldResetTimeout": false,
            "validateResponse": null,
            "retryCount": 0,
            "lastRequestTime": 1719906437941
        }
    },
    "request": {}
}`

File: src/containers/Tenant/Schema/SchemaTree/SchemaTree.tsx [hunk 17]
```diff
@@ -19,10 +23,12 @@ interface SchemaTreeProps {
 
 export function SchemaTree(props: SchemaTreeProps) {
     const {rootPath, rootName, rootType, currentPath, onActivePathUpdate} = props;
-
     const dispatch = useTypedDispatch();
 
     const [_, setQueryMode] = useQueryModes();
+    const [createDirectoryOpen, setCreateDirectoryOpen] = React.useState(false);
+    const [parentPath, setParentPath] = React.useState('');
+    const [path, setPath] = React.useState('');
```
Comment (ValeraS) [inline src/containers/Tenant/Schema/SchemaTree/SchemaTree.tsx:31]: this is used only as a key, but from name it's not clear.

File: src/containers/Tenant/Schema/CreateDirectoryDialog/CreateDirectoryDialog.tsx [hunk 18]
```diff
@@ -0,0 +1,96 @@
...
+}
+
+export function CreateDirectoryDialog({open, onClose, parentPath, onSuccess}: SchemaTreeProps) {
+    const [error, setError] = React.useState<unknown>('');
+    const [relativePath, setRelativePath] = React.useState('');
+    const [create, response] = schemaApi.useCreateDirectoryMutation();
+
+    const handleUpdate = (updated: string) => {
+        setRelativePath(updated);
+        setError(validateRelativePath(updated));
```
Comment (ValeraS) [inline src/containers/Tenant/Schema/CreateDirectoryDialog/CreateDirectoryDialog.tsx:35]: Do we really need to validate field on change?

**Review by ValeraS — CHANGES_REQUESTED** [review]

File: src/containers/Tenant/Schema/CreateDirectoryDialog/CreateDirectoryDialog.tsx [hunk 19]
```diff
@@ -0,0 +1,96 @@
...
+                    value={relativePath}
+                    onUpdate={handleUpdate}
+                    autoFocus
+                    hasClear
+                    disabled={response.isLoading}
+                    error={hasError}
+                />
+                <div className={b('error-wrapper')}>
+                    {hasError && (
+                        <ResponseError
```
Comment (ValeraS) [inline src/containers/Tenant/Schema/CreateDirectoryDialog/CreateDirectoryDialog.tsx:79]: Perhaps we need support for retrieving error messages from `error.data` https://github.com/ydb-platform/ydb-embedded-ui/pull/958#discussion_r1662010463

File: src/containers/Tenant/Schema/CreateDirectoryDialog/CreateDirectoryDialog.tsx (hunk 18 above)
Comment (Raubzeug) [reply src/containers/Tenant/Schema/CreateDirectoryDialog/CreateDirectoryDialog.tsx:35]: Actually no

File: src/containers/Tenant/Schema/CreateDirectoryDialog/CreateDirectoryDialog.tsx [hunk 20]
```diff
@@ -0,0 +1,116 @@
...
+                        placeholder={i18n('schema.tree.dialog.placeholder')}
+                        value={relativePath}
+                        onUpdate={handleUpdate}
+                        autoFocus
+                        hasClear
+                        disabled={response.isLoading}
+                        error={hasValidationError}
+                    />
+                    <div className={b('error-wrapper')}>
+                        {hasValidationError && <ResponseError error={validationError} />}
```
Comment (ValeraS) [inline src/containers/Tenant/Schema/CreateDirectoryDialog/CreateDirectoryDialog.tsx:97]: why is this not in the TextInput's errorMessage?

File: src/containers/Tenant/Schema/CreateDirectoryDialog/CreateDirectoryDialog.tsx [hunk 21]
```diff
@@ -0,0 +1,116 @@
...
+                            defaultMessage={i18n('schema.tree.dialog.invalid')}
+                        />
+                    )}
+                </Dialog.Body>
+                <Dialog.Footer
+                    loading={response.isLoading}
+                    textButtonApply={i18n('schema.tree.dialog.buttonApply')}
+                    textButtonCancel={i18n('schema.tree.dialog.buttonCancel')}
+                    onClickButtonCancel={handleClose}
+                    propsButtonApply={{disabled: hasValidationError, type: 'submit'}}
```
Comment (ValeraS) [inline src/containers/Tenant/Schema/CreateDirectoryDialog/CreateDirectoryDialog.tsx:111]:
```suggestion
                    propsButtonApply={{type: 'submit'}}
```

File: src/containers/Tenant/Schema/CreateDirectoryDialog/CreateDirectoryDialog.tsx [hunk 22]
```diff
@@ -0,0 +1,116 @@
...
+
+function validateRelativePath(value: string) {
+    if (value && /\s/.test(value)) {
+        return i18n('schema.tree.dialog.whitespace');
+    }
+    return '';
+}
+
+export function CreateDirectoryDialog({open, onClose, parentPath, onSuccess}: SchemaTreeProps) {
+    const [error, setError] = React.useState<unknown>('');
```
Comment (ValeraS) [inline src/containers/Tenant/Schema/CreateDirectoryDialog/CreateDirectoryDialog.tsx:29]: Why is this needed? `response` contains error.

File: src/containers/Tenant/Schema/CreateDirectoryDialog/CreateDirectoryDialog.tsx [hunk 23]
```diff
@@ -0,0 +1,116 @@
...
+                        <div>{`${parentPath}/`}</div>
+                    </div>
+                    <TextInput
+                        placeholder={i18n('schema.tree.dialog.placeholder')}
+                        value={relativePath}
+                        onUpdate={handleUpdate}
+                        autoFocus
+                        hasClear
+                        disabled={response.isLoading}
+                        error={hasValidationError}
```
Comment (ValeraS) [inline src/containers/Tenant/Schema/CreateDirectoryDialog/CreateDirectoryDialog.tsx:94]: error prop is deprecated.

**Review by ValeraS — CHANGES_REQUESTED** [review]

File: src/containers/Tenant/Schema/CreateDirectoryDialog/CreateDirectoryDialog.tsx (hunk 20 above)
Comment (Raubzeug) [reply src/containers/Tenant/Schema/CreateDirectoryDialog/CreateDirectoryDialog.tsx:97]: To prevent dialog's height jumping.

File: src/containers/Tenant/Schema/CreateDirectoryDialog/CreateDirectoryDialog.tsx (hunk 21 above)
Comment (Raubzeug) [reply src/containers/Tenant/Schema/CreateDirectoryDialog/CreateDirectoryDialog.tsx:111]: You suggest not to disable Apply button at all? But why?

Comment (ValeraS) [reply src/containers/Tenant/Schema/CreateDirectoryDialog/CreateDirectoryDialog.tsx:111]: Why do you want to disable it? There is a validation before submission.

Comment (Raubzeug) [reply src/containers/Tenant/Schema/CreateDirectoryDialog/CreateDirectoryDialog.tsx:111]: If it's validation error, we 100% know, that "Apply" won't work.
Or maybe I just like to forbid :)

File: src/containers/Tenant/Schema/CreateDirectoryDialog/CreateDirectoryDialog.tsx [hunk 24]
```diff
@@ -26,14 +28,14 @@ function validateRelativePath(value: string) {
 }
 
 export function CreateDirectoryDialog({open, onClose, parentPath, onSuccess}: SchemaTreeProps) {
-    const [error, setError] = React.useState<unknown>('');
+    const [showResponseError, setShowResponseError] = React.useState(false);
```
Comment (ValeraS) [inline src/containers/Tenant/Schema/CreateDirectoryDialog/CreateDirectoryDialog.tsx:31]: `response` has a `reset` method

File: src/containers/Tenant/Schema/CreateDirectoryDialog/CreateDirectoryDialog.tsx [hunk 25]
```diff
@@ -78,27 +79,29 @@ export function CreateDirectoryDialog({open, onClose, parentPath, onSuccess}: Sc
                 }}
             >
                 <Dialog.Body>
-                    <div className={b('label')}>
-                        <div className={b('description')}>
+                    <label htmlFor={relativePathInputName} className={b('label')}>
```
Comment (ValeraS) [inline src/containers/Tenant/Schema/CreateDirectoryDialog/CreateDirectoryDialog.tsx:82]: `for` attribute should be connected to an `id` of input not a `name`.

**Review by ValeraS — CHANGES_REQUESTED** [review]

**Review by ValeraS — APPROVED** [review]

**Review by Raubzeug — APPROVED** [review]
